lib/websocket/connection.js

```diff
@@ -0,0 +1,365 @@
+      //    subprotocol not requested by the client), the client MUST _Fail
+      //    the WebSocket Connection_.
+      const secProtocol = response.headersList.get('Sec-WebSocket-Protocol')
+
+      if (secProtocol !== null && secProtocol !== request.headersList.get('Sec-WebSocket-Protocol')) {
+        failWebsocketConnection(controller, response.socket)
+        return
+      }
+
+      // processResponse is called when the "response’s header list has been received and initialized."
+      // once this happens, the connection is open
+      ws[kResponse] = response
+
+      whenConnectionEstablished(ws)
+
+      receiveData(ws)
+
+      socketClosed(ws)
+    }
+  })
+
+  return controller
+}
+
+/**
+ * @param {import('../fetch/index').Fetch} controller
+ * @param {import('net').Socket|import('tls').TLSSocket|undefined} socket
+ */
+function failWebsocketConnection (controller, socket) {
+  controller.abort()
+
+  if (socket && !socket.destroyed) {
+    socket.destroy()
+  }
+}
+
+/**
+ * @see https://websockets.spec.whatwg.org/#feedback-from-the-protocol
+ * @param {import('./websocket').WebSocket} ws
+ */
+function whenConnectionEstablished (ws) {
+  const { [kResponse]: response } = ws
+
+  // 1. Change the ready state to OPEN (1).
+  ws[kReadyState] = states.OPEN
+
+  // 2. Change the extensions attribute’s value to the extensions in use, if
+  //    it is not the null value.
+  // https://datatracker.ietf.org/doc/html/rfc6455#section-9.1
+  const extensions = response.headersList.get('sec-websocket-extensions')
+
+  if (extensions !== null) {
+    ws[kExtensions] = extensions
+  }
+
+  // 3. Change the protocol attribute’s value to the subprotocol in use, if
+  //    it is not the null value.
+  // https://datatracker.ietf.org/doc/html/rfc6455#section-1.9
+  const protocol = response.headersList.get('sec-websocket-protocol')
+
+  if (protocol !== null) {
+    ws[kProtocol] = protocol
+  }
+
+  // 4. Fire an event named open at the WebSocket object.
+  fireEvent('open', ws)
+}
+
+/**
+ * @see https://websockets.spec.whatwg.org/#feedback-from-the-protocol
+ * @param {import('./websocket').WebSocket} ws
+ */
+function receiveData (ws) {
+  const { [kResponse]: response } = ws
+
+  /** @type {WebsocketFrame|undefined} */
+  let frame
+
+  response.socket.on('data', (chunk) => {
+    if (!frame) {
+      frame = WebsocketFrame.from(chunk)
+      frame.terminated = frame.fin // message complete
+    } else {
+      // A fragmented message consists of a single frame with the FIN bit
+      // clear and an opcode other than 0, followed by zero or more frames
+      // with the FIN bit clear and the opcode set to 0, and terminated by
+      // a single frame with the FIN bit set and an opcode of 0.
+
+      const fragmentedFrame = WebsocketFrame.from(chunk)
+
+      if (fragmentedFrame.fin && fragmentedFrame.opcode === opcodes.CONTINUATION) {
+        frame.terminated = true
+      }
+
+      frame.addFrame(fragmentedFrame.data)
+    }
```

Comment:
At a glance this does not handle the case where a control frame (ping/pong) is received in the middle of a fragmented message. If it is, ignore this comment.

Reply:
I had no idea this could happen, I will fix it.

Reply:
As per https://github.com/KhafraDev/undici/pull/30#discussion_r1039009237, another problem I see is that the WHATWG spec advertises support for the `permessage-deflate` extension. If the server you are connecting to also supports `permessage-deflate`, you might receive frames whose payload data is compressed.

Reply:
I don't know if I'll implement permessage deflate support in this pr; the ping/pong issue has been fixed.

Reply:
Ok, then I think it would be better to not send the `Sec-WebSocket-Extensions` header. The connection would be surprisingly closed every time the server sends a compressed message. Plus there is currently no validation and negotiation of the server sent `Sec-WebSocket-Extensions` header. If you enable `permessage-deflate` on the server, some of web-platform-tests fails.

Reply:
I've disabled it